### PR TITLE
cl/ssz, common/ssz, engineapi: add recursive SSZ offset and length validation

### PR DESCRIPTION
**Pull request by yperbasis**
## Summary

- add opt-in validation for SSZ container and list offsets, encoding lengths, and boolean encodings
- reject gaps between fixed and dynamic sections, malformed dynamic-list offset tables, oversized extra data, non-canonical boolean encodings, and trailing bytes in all-static schemas
- propagate validation through nested lists, execution payloads, and payload attributes
- fall back to permissive nested decoding when a type has no strict decoder, with coverage tests guarding every current Engine schema fallback
- correctly decode the canonical four-byte representation of a single empty element on permissive dynamic-list paths
- keep Electra and Fulu `PayloadAttributes` on the Cancun-compatible wire shape; add `slot_number` and `target_gas_limit` only from Gloas
- keep strict decoding opt-in at existing decoder call sites

## Motivation

SSZ offsets define the canonical boundary between fixed and variable sections, and a fixed-size type has exactly one encoding length. A decoder that validates only the outer container can still accept unreferenced gap bytes or trailing padding inside nested values. SSZ booleans likewise have only two canonical byte encodings.

#21729 is the first consumer of these strict APIs. Its Engine API REST-SSZ request decoders require canonical offsets and encoding lengths throughout their nested request schemas. This PR provides that reusable validation while keeping strict decoding opt-in.

## Testing

- `go test ./common/ssz ./cl/ssz/... ./cl/cltypes ./cl/cltypes/solid ./execution/engineapi/engine_types -count=1`
- `go test ./execution/engineapi/... -count=1`
- `make lint` (repeated)
- `make erigon integration`

### CODE REVIEW

File: cl/ssz/decode.go [hunk 1]
```diff
@@ -62,58 +51,57 @@ func UnmarshalSSZ(buf []byte, version int, schema ...any) (err error) {
...
 				if len(buf) < position+obj.EncodingSizeSSZ() {
 					return ssz.ErrLowBufferSize
 				}
-				// If the object is static (fixed size), decode it from the buf and update the position
-				if err = obj.DecodeSSZ(buf[position:], version); err != nil {
+				if strict {
+					err = decodeObjectSSZStrict(obj, buf[position:], version)
+				} else {
+					err = obj.DecodeSSZ(buf[position:], version)
+				}
```
[the remark this answers is not in the document]
Comment (yperbasis) [reply cl/ssz/decode.go:94]: Thanks. Strict decoding in this PR is intentionally limited to canonical offset validation. Under the SSZ type rules, an object for which `Static()` returns true cannot contain variable-size sub-elements, so it has no nested offsets to validate. All current `StrictUnmarshaler` implementations used through `SizedObjectSSZ` are variable-size.

Passing an exact slice would be a separate scope-boundary cleanup, and changing the permissive call would also alter established decoder behavior. If strict decoding is expanded beyond offsets, that should be handled and tested separately. I am therefore leaving the static decoding behavior unchanged here.

**Comment by AskAlexSharov** [comment]
## Review

Design is sound: opt-in, existing callers untouched, and I could not make strict reject anything canonical. Verified locally on `82e1ad2`:

- `go test ./common/ssz/... ./cl/ssz/... ./cl/cltypes/... ./execution/engineapi/... -count=1` — all green.
- Throwaway round-trip tests: `ExecutionPayload.DecodeSSZStrict` accepts canonically-encoded payloads for Bellatrix/Capella/Deneb/Electra/Fulu **and Gloas** (including empty tx + withdrawal lists, and a single *empty* transaction), and `PayloadAttributes.DecodeSSZStrict` likewise for Gloas. No false rejections.
- Re-derived the offset logic in `unmarshalSSZ`: dynamic element ranges are contiguous by construction (`endOffset` of element *i* is the start of *i+1*), so `offsets[0] != position` plus the pre-existing `offsets[i] > endOffset` check is sufficient — no gap or overlap can slip past. Likewise `TransactionsSSZ.DecodeSSZ` is already canonical-strict on its own (length derived from `firstOffset`, `startTx >= firstOffset`, monotonic ends, final end pinned to `len(buf)`), so its missing `DecodeSSZStrict` is not a hole today.

Five things below; the first two are worth acting on.

### 1. Strict and lax `DecodeDynamicList` return different *values*, not just different errors

```go
func (e *divElem) DecodeSSZ(buf []byte, _ int) error { e.n = len(buf); return nil }
func (e *divElem) Clone() clonable.Clonable          { return &divElem{} }

buf := []byte{4, 0, 0, 0} // canonical List[T,N] with one element whose encoding is empty

lax, _    := DecodeDynamicList[*divElem](buf, 0, 4, 8, 0)       // len(lax)    == 0
strict, _ := DecodeDynamicListStrict[*divElem](buf, 0, 4, 8, 0) // len(strict) == 1
```

The lax bound `len(buf) > 4` should be `>= 4`. A canonical list with exactly one variable-size element whose encoding is empty is 4 bytes (offset table `[4]`, no payload) — lax drops the element, strict keeps it. Same shape for `len(buf)` in 1..3: lax silently returns an empty list, strict returns `ErrLowBufferSize`.

So `DecodeSSZStrict` is not "`DecodeSSZ` plus validation" — on some inputs it decodes to something else. Strict is the spec-correct side, and the lax side is a real round-trip bug (`ListSSZ.EncodeSSZ` of one empty element does not survive `ListSSZ.DecodeSSZ`). Either fix the bound here or state the divergence explicitly, because right now the two paths quietly disagree and nothing flags it.

### 2. Strict propagation fails open

`decodeObjectStrict` / `decodeObjectSSZStrict` fall back to `DecodeSSZ` for any type that doesn't implement `StrictUnmarshaler`. Exactly four production types implement it: `*ExecutionPayload`, `*PayloadAttributes`, `*Eth1Block`, `*ListSSZ[T]`. Everything else reachable from those schemas — `*TransactionsSSZ`, `*ByteListSSZ`, `*cltypes.Withdrawal` — takes the permissive path.

That's fine today (checked above), but it's the classic fail-open shape: someone adds a variable-size field to `Eth1Block`, forgets `DecodeSSZStrict`, and the guarantee #21729 depends on silently narrows with no test going red. Two cheap mitigations:

- note in the `UnmarshalSSZStrict` doc comment that strictness is best-effort per nested type — the current wording ("propagates strict decoding to nested objects that support it") is accurate but easy to read as a guarantee;
- add a test asserting that every `SizedObjectSSZ` in `Eth1Block.getSchema()` and the `PayloadAttributes` schemas either implements `StrictUnmarshaler` or sits on an explicit allowlist of "no internal offset table". ~15 lines, and it turns the invariant into something CI enforces instead of something a reviewer has to re-derive.

### 3. Duplicate helper

`common/ssz.decodeObjectStrict` and `cl/ssz.decodeObjectSSZStrict` are the same five lines. The `cl/ssz` one only needs `DecodeSSZ` off the value, so a single exported helper in `common/ssz` over a minimal interface would drop the copy.

### 4. `unmarshal := ssz2.UnmarshalSSZ`

In `PayloadAttributes.decodeSSZ`, dispatching through a func variable blocks inlining and devirtualization at all four call sites. Negligible at FCU rates, so purely a preference — but the same pattern has cost us in hot paths before, and threading `strict` down would read the same.

### 5. Test coverage

- No test pins the "canonical stays canonical" direction. A table test over all `StateVersion`s asserting `DecodeSSZStrict` accepts `EncodeSSZ` output is the one that guards against a false rejection sneaking into the still-moving Gloas schema. I ran it ad hoc and it passes — worth keeping in tree.
- No test for descending offsets between two dynamic fields (`offsets[i] > endOffset`) under strict.

Separately: deleting the algorithm-narrating block comment on `UnmarshalSSZ` is a good change on its own.

**Review by AskAlexSharov — APPROVED** [review]

File: common/ssz/ssz.go [hunk 2]
```diff
@@ -85,12 +100,33 @@ func UnmarshalUint64SSZ(x []byte) uint64 {
...
+	if strict && len(buf) != 0 {
+		if len(buf) < 4 {
+			return nil, ErrLowBufferSize
+		}
+		currentOffset = DecodeOffset(buf)
+		if currentOffset == 0 || currentOffset%4 != 0 || currentOffset > uint32(len(buf)) {
+			return nil, ErrBadOffset
+		}
+		elementsNum = currentOffset / 4
+	} else if len(buf) > 4 {
```
Comment (domiwei) [inline common/ssz/ssz.go]: Could we make this `len(buf) >= 4` and add a round-trip test for a single empty dynamic element? `{4, 0, 0, 0}` is the canonical encoding of one variable-size element whose encoding is empty. The strict path returns one element here, while the permissive path returns zero, so opting into strict decoding changes the decoded value for valid input rather than only rejecting malformed encodings.

Comment (yperbasis) [reply common/ssz/ssz.go]: Good catch. EncodeDynamicList emits {4, 0, 0, 0} for one empty dynamic element, so permissive decoding must handle a four-byte offset table. I changed the condition to len(buf) >= 4 and added a round-trip test covering both permissive and strict decoding in 02b245d006.
